java-vtl-script/src/main/java/no/ssb/vtl/script/operations/join/AbstractJoinOperation.java

```diff
@@ -85,51 +85,53 @@
 
         checkNotNull(identifiers);
 
-        this.joinScope = createJoinScope(namedDatasets);
-
-        this.componentMapping = createComponentMapping(this.datasets.values());
-
-        Map<Component, Map<Dataset, Component>> idMap = this.componentMapping.rowMap().entrySet().stream()
-                .filter(e -> e.getKey().isIdentifier())
-                .filter(e -> e.getValue().size() == datasets.size())
-                // identifiers can be from any dataset
-                .filter(e -> identifiers.isEmpty() || !Collections.disjoint(e.getValue().values(), identifiers))
-                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+        this.commonIdentifiers = ImmutableMap.copyOf(new CommonIdentifierBindings(this.datasets)
+                .getComponentReferences().entrySet().stream()
+                .filter(entry -> identifiers.isEmpty() || identifiers.contains(entry.getValue()))
+                .collect(Collectors.toSet()));
 
         // No common identifier
-        checkArgument(namedDatasets.size() == 1 || !idMap.isEmpty(), ERROR_NO_COMMON_IDENTIFIERS, namedDatasets);
+        checkArgument(namedDatasets.size() == 1 || !commonIdentifiers.isEmpty(), ERROR_NO_COMMON_IDENTIFIERS, namedDatasets);
 
-        this.commonIdentifiers = ImmutableSet.copyOf(idMap.keySet());
 
         // Type mismatch check
         List<String> typeMismatches = Lists.newArrayList();
-        for (Map.Entry<Component, Map<Dataset, Component>> entry : idMap.entrySet()) {
-            Component identifier = entry.getKey();
-
+        for (Map.Entry<String, Component> entry : commonIdentifiers.entrySet()) {
             Multimap<Class<?>, Dataset> typeMap = ArrayListMultimap.create();
-            entry.getValue().entrySet().forEach(datasetComponent -> {
-                typeMap.put(
-                        datasetComponent.getValue().getType(),
-                        datasetComponent.getKey()
-                );
-            });
-            if (typeMap.keySet().size() != 1)
-                typeMismatches.add(String.format("%s -> (%s)", identifier, typeMap));
+            for (Dataset dataset : datasets.values()) {
+
+                typeMap.put(dataset.getDataStructure().get(entry.getKey()).getType(), dataset);
+
+                if (typeMap.keySet().size() != 1)
+                    typeMismatches.add(String.format("%s -> (%s)", entry.getKey(), typeMap));
+            }
         }
         checkArgument(
                 typeMismatches.isEmpty(),
                 ERROR_INCOMPATIBLE_TYPES,
                 String.join(", ", typeMismatches)
         );
+        this.commonIdentifierNames = commonIdentifiers.keySet();
```

Comment:
Maybe use `commonIdentifiers.keySet()` instead of `commonIdentifierNames` so that we avoid confusion? I think having them in a Map<String, Component> is very mush appropriate.

Reply:
Agree. I'll use commonIdentifiers.keySet() instead.